src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -106,215 +125,272 @@ class column_pruner {
      */
     column_pruner(const std::vector<std::string> & names, 
             const std::vector<int> & num_children, 
-            int parent_num_children): children(), s_id(0), c_id(-1) {
-      add_depth_first(names, num_children, parent_num_children);
+            const std::vector<int> & tags, 
+            int parent_num_children): children(), tag(STRUCT_TAG) {
+      add_depth_first(names, num_children, tags, parent_num_children);
+    }
+
+    column_pruner(int in_tag): children(), tag(in_tag) {
+    }
+
+    column_pruner(): children(), tag(STRUCT_TAG) {
+    }
+
+    void skip(std::vector<parquet::format::SchemaElement> & schema,
+            std::size_t & current_input_schema_index, std::size_t & next_input_chunk_index) {
+      // We want to skip everything referenced by the current_input_schema_index and its children.
+      // But we do have to update the chunk indexes as we go.
+      int num_to_skip = 1;
+      while(num_to_skip > 0 && current_input_schema_index < schema.size()) {
+        auto schema_item = schema[current_input_schema_index];
+        bool is_leaf = schema_item.__isset.type;
+        if (is_leaf) {
+          ++next_input_chunk_index;
+        }
+
+        if (schema_item.__isset.num_children) {
+          num_to_skip = num_to_skip + schema_item.num_children;
+        }
+
+        --num_to_skip;
+        ++current_input_schema_index;
+      }
     }
 
-    column_pruner(int s_id, int c_id): children(), s_id(s_id), c_id(c_id) {
+    std::string get_name(parquet::format::SchemaElement & elem, const bool normalize_case = false) {
+      if (normalize_case) {
+        return unicode_to_lower(elem.name);
+      } else {
+        return elem.name;
+      }
     }
 
-    column_pruner(): children(), s_id(0), c_id(-1) {
+    column_pruner * find_child(std::string name) {
+      auto found_it = children.find(name);
+      if (found_it != children.end()) {
+        return &(found_it->second);
+      }
+      return nullptr;
+    }
+
+    int get_num_children(parquet::format::SchemaElement & elem) {
+      if (elem.__isset.num_children) {
+        return elem.num_children;
+      }
+      return 0;
+    }
+
+    parquet::format::SchemaElement & get_schema_item(std::vector<parquet::format::SchemaElement> & schema, const std::size_t index) {
+      if (index >= schema.size()) {
+        throw std::runtime_error("walked off the end of the schema some how...");
+      }
+      return schema[index];
```

Comment:
The entire of this function is not necessary. If you want to access with bound check then just use `schema.get(index)` at the caller. If you want to have some alias function like this then just:
```suggestion
    parquet::format::SchemaElement & get_schema_item(std::vector<parquet::format::SchemaElement> & schema, const std::size_t index) {
      return schema.get(index);
```